mlflow/models/evaluation/base.py

```diff
@@ -151,6 +156,170 @@ def __str__(self):
         return "EvaluationMetric(" + ", ".join(parts) + ")"
 
 
+# NB: we need this function because we cannot modify the signature of
+# a class's __call__ method after the class has been defined.
+# This is also useful to distinguish between the metric signatures with different eval_fn signatures
+def dynamically_generate_eval_metric(eval_fn, standard_signature=False):
+    """
+    Dynamically generate a GenAIEvaluationMetric class that can be used to evaluate the metric
+    on the given input data. The generated class is callable with a __call__ method that
+    takes the arguments specified in the signature of the eval_fn function.
+
+    Args:
+        eval_fn: the evaluation function of the EvaluationMetric.
+        standard_signature: (Optional) Whether the metric follows a strict signature, if True then
+            the eval_fn must follow below signature:
```

Comment:
what would happen if it doesn't?

Reply:
Errors would happen if the underlying eval_fn doesn't accept positional args as we passed in, and if any required arguments are not passed / extra arguments passed we also raise error.
If the invocation order in the underlying model is different (e.g. def eval_fn(inputs, predictions, metrics..)) then the error would be harder to find.
This should only be used internally so developers should make sure we don't mess it up.
But I think Yuki will refactor them at some point to unify these two eval_fn signatures (breaking change) so we won't need to worry after that, and ideally the eval_fn should only accept keyword arguments to avoid any issues.